File: OC/TabbarFrameDemo/UI/MainTabbarController/JCMainTabbarController.m
//
//  JCMainTabbarController.m
//  TabbarFrameDemo
//
//  Created by jiechen on 15/9/13.
//  Copyright © 2015年 jch. All rights reserved.
//

#import "JCMainTabbarController.h"
#import "JCHomePageViewController.h"
#import "JCFindViewController.h"
#import "JCMyCenterViewController.h"

@interface JCMainTabbarController ()

@end

@implementation JCMainTabbarController

+ (instancetype)sharedInstance
{
    static JCMainTabbarController *tabbarVC = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        tabbarVC = [[JCMainTabbarController alloc]init];
    });
    return tabbarVC;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    [self loadSubviewControllers];
}

- (void)loadSubviewControllers
{
    //子控制器
    NSArray *viewControllers = @[[JCHomePageViewController getInstance],
                                 [JCFindViewController getInstance],
                                 [JCMyCenterViewController getInstance]];
    
    //tabBarItem的标题，如果不配置，则使用子控制器的标题。
    NSArray *tabTitles = @[@"首页", @"发现", @"我的"];
    NSArray *tabImages = @[[UIImage imageNamed:@"tabbar_myCenter_nor"], //目前没有素材
                           [UIImage imageNamed:@"tabbar_find_nor"],
                           [UIImage imageNamed:@"tabbar_myCenter_nor"]];
    NSArray *tabSelectedImages = @[];
    
    NSMutableArray *navigationControllers = [NSMutableArray arrayWithCapacity:viewControllers.count];
    [viewControllers enumerateObjectsUsingBlock:^(UIViewController *  _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
        UINavigationController *navCon =[[UINavigationController alloc] initWithRootViewController:obj];
        navCon.tabBarItem.title = tabTitles[idx];
        navCon.tabBarItem.image = tabImages[idx];
        [navigationControllers addObject:navCon];
    }];
    self.viewControllers = navigationControllers;
}

@end

File: Swift/TabbarFrameDemo/JCMainTabbarController.swift
//
//  JCMainTabbarController.swift
//  TabbarFrameDemo
//
//  Created by jiechen on 15/11/10.
//  Copyright © 2015年 jiechen. All rights reserved.
//

import UIKit

class JCMainTabbarController: UITabBarController {

    override func viewDidLoad() {
        super.viewDidLoad()

        // Do any additional setup after loading the view.
//        self.configSystemTabbar()
        self.configCustomTabbar()
    }
    
    func configSystemTabbar() {
        let firstController = FirstViewController()
        let secondController = SecondViewController()
        let thirdController = ThirdViewController()
        
        firstController.tabBarItem = UITabBarItem(tabBarSystemItem: UITabBarSystemItem.MostRecent, tag: 0)
        firstController.view.backgroundColor = UIColor.redColor()
        
        secondController.tabBarItem = UITabBarItem(tabBarSystemItem: UITabBarSystemItem.Contacts, tag: 1)
        secondController.view.backgroundColor = UIColor.lightGrayColor()
        
        thirdController.tabBarItem = UITabBarItem(tabBarSystemItem: UITabBarSystemItem.Bookmarks, tag: 2)
        thirdController.view.backgroundColor = UIColor.blueColor()
        
        self.viewControllers = [firstController, secondController, thirdController];
    }
    
    func configCustomTabbar() {
        let controllers = [FirstViewController(), SecondViewController(), ThirdViewController()]
        let tabTitles = ["首页", "发现", "我的"]
        let tabImages = [UIImage(named: "tabbar_myCenter_nor"), UIImage(named: "tabbar_find_nor"), UIImage(named: "tabbar_myCenter_nor")]
        
        var navControllers: [UINavigationController] = []
        for idx in 0..<controllers.count {
            let navController = UINavigationController(rootViewController: controllers[idx])
            navController.tabBarItem.title = tabTitles[idx]
            navController.tabBarItem.image = tabImages[idx]
            navControllers.append(navController)
        }
        self.viewControllers = navControllers
    }
}
